core/src/main/java/io/questdb/griffin/SqlOptimiser.java

```diff
@@ -4557,31 +4560,43 @@ private QueryModel rewriteSampleBy(@Nullable QueryModel model) throws SqlExcepti
         QueryModel nested = model.getNestedModel();
         if (nested != null) {
             // "sample by" details will be on the nested model of the query
-            ExpressionNode sampleBy = nested.getSampleBy();
-            ExpressionNode sampleByOffset = nested.getSampleByOffset();
-            ObjList<ExpressionNode> sampleByFill = nested.getSampleByFill();
-            ExpressionNode sampleByTimezoneName = nested.getSampleByTimezoneName();
-            ExpressionNode sampleByUnit = nested.getSampleByUnit();
-            ExpressionNode timestamp = nested.getTimestamp();
+            final ExpressionNode sampleBy = nested.getSampleBy();
+            final ExpressionNode sampleByOffset = nested.getSampleByOffset();
+            final ObjList<ExpressionNode> sampleByFill = nested.getSampleByFill();
+            final ExpressionNode sampleByTimezoneName = nested.getSampleByTimezoneName();
+            final ExpressionNode sampleByUnit = nested.getSampleByUnit();
+            final ExpressionNode timestamp = nested.getTimestamp();
+            final int sampleByFillSize = sampleByFill.size();
+            ExpressionNode sampleByFrom = nested.getSampleByFrom();
+            ExpressionNode sampleByTo = nested.getSampleByTo();
 
             if (
                     sampleBy != null
                             && timestamp != null
                             && (sampleByOffset != null && SqlKeywords.isZeroOffset(sampleByOffset.token) && (sampleByTimezoneName == null || SqlKeywords.isUTC(sampleByTimezoneName.token)))
-                            && (sampleByFill.size() == 0 || (sampleByFill.size() == 1 && SqlKeywords.isNoneKeyword(sampleByFill.getQuick(0).token)))
+                            && (sampleByFillSize == 0 || (sampleByFillSize == 1 && !SqlKeywords.isPrevKeyword(sampleByFill.getQuick(0).token) && !SqlKeywords.isLinearKeyword(sampleByFill.getQuick(0).token)))
                             && sampleByUnit == null
+                            && !((sampleByFrom != null && sampleByFrom.token.charAt(0) == '$') || (sampleByTo != null && sampleByTo.token.charAt(0) == '$'))
```

Comment:
What's the purpose of this check? It looks weird + will lead to index out of bounds exception in case if any of the tokens is an empty string.

Reply:
We have 3 syntaxes for bind variables: `?` (PGWire), `$<num>` (embedded, indexed) and `:<name>` (embedded, named), so the check is incomplete. There is a special `ExpressionNode.BIND_VARIABLE` constant for the bind variable type, yet there is a catch: the expression may be something like `abs(?)` or `? + ?`, i.e. have a wrapping function and/or expression around bind variables, so a tree traversal may be needed to learn if there are any bind variables.